Add support for a new option, "--almost-all-if-not-home", that implicitly shows
dotfiles for every directory except `$HOME`.

--- src/ls.c	2023-08-29 04:39:27.000000000 -0700
+++ src/ls.c	2023-11-18 19:33:32.904591051 -0800
@@ -261,7 +261,7 @@
                               int needs_general_quoting, size_t *width,
                               bool *pad);
 static int decode_switches (int argc, char **argv);
-static bool file_ignored (char const *name);
+static bool file_ignored (bool is_home, char const *name);
 static uintmax_t gobble_file (char const *name, enum filetype type,
                               ino_t inode, bool command_line_arg,
                               char const *dirname);
@@ -382,6 +382,10 @@
 
 static struct pending *pending_dirs;
 
+// This is used to store the stat information of the user's home directory
+// which will be used to determine when its contents are being listed.
+static struct stat homedir_stat;
+
 /* Current time in seconds and nanoseconds since 1970, updated as
    needed when deciding whether a file is recent.  */
 
@@ -843,6 +847,7 @@
   TIME_OPTION,
   TIME_STYLE_OPTION,
   ZERO_OPTION,
+  ALMOST_ALL_IF_NOT_HOME
 };
 
 static struct option const long_options[] =
@@ -891,6 +896,7 @@
   {"block-size", required_argument, nullptr, BLOCK_SIZE_OPTION},
   {"context", no_argument, 0, 'Z'},
   {"author", no_argument, nullptr, AUTHOR_OPTION},
+  {"almost-all-if-not-home", no_argument, NULL, ALMOST_ALL_IF_NOT_HOME},
   {GETOPT_HELP_OPTION_DECL},
   {GETOPT_VERSION_OPTION_DECL},
   {nullptr, 0, nullptr, 0}
@@ -958,6 +964,8 @@
   size_t *col_arr;
 };
 
+static bool almost_all_if_not_home;
+
 /* Array with information about column fullness.  */
 static struct column_info *column_info;
 
@@ -1666,6 +1674,8 @@
   current_time.tv_sec = TYPE_MINIMUM (time_t);
   current_time.tv_nsec = -1;
 
+  almost_all_if_not_home = false;
+
   i = decode_switches (argc, argv);
 
   if (print_with_color)
@@ -1784,6 +1794,9 @@
   else if (n_files <= 1 && pending_dirs && pending_dirs->next == 0)
     print_dir_name = false;
 
+  if (!almost_all_if_not_home || !getenv("HOME") || stat(getenv("HOME"), &homedir_stat) < 0)
+    homedir_stat.st_dev = homedir_stat.st_ino = 0;
+
   while (pending_dirs)
     {
       thispend = pending_dirs;
@@ -2233,6 +2246,10 @@
           quoting_style_opt = literal_quoting_style;
           break;
 
+        case ALMOST_ALL_IF_NOT_HOME:
+          almost_all_if_not_home = true;
+          break;
+
         case_GETOPT_HELP_CHAR;
 
         case_GETOPT_VERSION_CHAR (PROGRAM_NAME, AUTHORS);
@@ -2993,22 +3010,25 @@
       return;
     }
 
-  if (LOOP_DETECT)
-    {
-      struct stat dir_stat;
-      int fd = dirfd (dirp);
+  struct stat dir_stat;
+  int fd = dirfd (dirp);
 
-      /* If dirfd failed, endure the overhead of stat'ing by path  */
-      if ((0 <= fd
-           ? fstat_for_ino (fd, &dir_stat)
-           : stat_for_ino (name, &dir_stat)) < 0)
-        {
-          file_failure (command_line_arg,
-                        _("cannot determine device and inode of %s"), name);
-          closedir (dirp);
-          return;
-        }
+  /* If dirfd failed, endure the overhead of stat'ing by path  */
+  if ((0 <= fd
+       ? fstat_for_ino (fd, &dir_stat)
+       : stat_for_ino (name, &dir_stat)) < 0)
+    {
+      file_failure (command_line_arg,
+                    _("cannot determine device and inode of %s"), name);
+      closedir (dirp);
+      return;
+    }
+
+  bool is_home = (dir_stat.st_ino == homedir_stat.st_ino &&
+                  dir_stat.st_dev == homedir_stat.st_dev);
 
+  if (LOOP_DETECT)
+    {
       /* If we've already visited this dev/inode pair, warn that
          we've found a loop, and do not process this directory.  */
       if (visit_dir (dir_stat.st_dev, dir_stat.st_ino))
@@ -3059,7 +3079,7 @@
       next = readdir (dirp);
       if (next)
         {
-          if (! file_ignored (next->d_name))
+          if (! file_ignored (is_home, next->d_name))
             {
               enum filetype type = unknown;
 
@@ -3175,11 +3195,11 @@
 /* Return true if FILE should be ignored.  */
 
 static bool
-file_ignored (char const *name)
+file_ignored (bool is_home, char const *name)
 {
   return ((ignore_mode != IGNORE_MINIMAL
            && name[0] == '.'
-           && (ignore_mode == IGNORE_DEFAULT || ! name[1 + (name[1] == '.')]))
+           && (ignore_mode == IGNORE_DEFAULT && (!almost_all_if_not_home || is_home || ! name[1 + (name[1] == '.')])))
           || (ignore_mode == IGNORE_DEFAULT
               && patterns_match (hide_patterns, name))
           || patterns_match (ignore_patterns, name));
